waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,605 @@
+        make-kill-response (fn [killed? message status]
+                             {:instance-id id :killed? killed?
+                              :message message :service-id service-id :status status})]
+    ; request termination of the instance
+    (api-request http-client pod-url :request-method :delete :body term-json)
+    ; scale down the replicaset to reflect removal of this instance
+    (scale-service-by-delta scheduler service -1)
+    ; force-kill the instance (should still be terminating)
+    (api-request http-client pod-url :request-method :delete :body kill-json)
+    ; report back that the instance was killed successfully
+    (scheduler/process-instance-killed! instance)
+    200))
+
+(defn- service-spec
+  "Creates a Kubernetes ReplicaSet spec (with an embedded Pod spec) for the given Waiter Service."
+  [{:keys [rs-spec-file-path pod-base-port] :as scheduler} service-id service-description service-id->password-fn]
+  (let [{:strs [backend-proto cmd cpus grace-period-secs health-check-interval-secs
+                health-check-max-consecutive-failures mem min-instances ports run-as-user]} service-description
+        home-path (str "/home/" run-as-user)
+        common-env (scheduler/environment service-id service-description
+                                          service-id->password-fn home-path)
+        port0 pod-base-port
+        template-env (into [;; We set these two "MESOS_*" variables to improve interoperability
+                            {:name "MESOS_DIRECTORY" :value home-path}
+                            {:name "MESOS_SANDBOX" :value home-path}]
+                           (concat
+                             (for [[k v] common-env]
+                               {:name k :value v})
+                             (for [i (range ports)]
+                               {:name (str "PORT" i) :value (str (+ port0 i))})))
+        params {:k8s-name (service-id->k8s-name scheduler service-id)
+                :backend-protocol backend-proto
+                :backend-protocol-caps (string/upper-case backend-proto)
+                :cmd cmd
+                :cpus cpus
+                :env template-env
+                :grace-period-secs grace-period-secs
+                :health-check-interval-secs health-check-interval-secs
+                :health-check-max-consecutive-failures health-check-max-consecutive-failures
+                :health-check-url (sd/service-description->health-check-url service-description)
+                :home-path home-path
+                :memory  (str mem "Mi")
+                :min-instances min-instances
+                :port-count ports
+                :run-as-user run-as-user
+                :service-id service-id
+                :ssl? (= "https" backend-proto)}
+        edn-opts {:readers {'waiter/param params
+                            'waiter/param-str (comp str params)
+                            'waiter/port #(+ port0 %)
```

Comment:
Should port also be exposed as params `PORT0`, `PORT1` and so on based on number of requested ports?

Reply:
I think it's already doing what you're requesting. E.g., in the default ReplicaSet spec:
```clojure
:ports [{:containerPort #waiter/port 0}]
```
`#waiter/port 0` above could be replaced with `#waiter/port 5` or any other valid number.

Reply:
I meant refer to it as `#waiter/param :PORT0`

Reply:
Is there an advantage to writing `#waiter/param :PORT0` vs `#waiter/port 0`? Is the former less confusing? I feel like we should pick just one or the other.

Reply:
I think I am more familiar with what `PORT0` means, whereas I had to lookup the code to understand what `#waiter/port 0` was doing.

Reply:
All of the other `param` values are lower-case. What about `#waiter/env PORT0`? I.e., we'd be exposing all of the values from the provided environment (including `PORT0` – `PORT9`, if available) for the user to reference inside the spec.

Reply:
It might also make sense to sanitize the arguments to the `param` and `env` reader functions (e.g., pass them thru `keyword` or `name`) so symbols/strings/keywords all work sensibly.

Reply:
I just thought of an issue with my `#waiter/env` idea: all environment values are *strings*, but we need an *integer* for this use case. There's no reason that `#waiter/param :port0` couldn't return an integer, but the `#waiter/env` idea doesn't really work here.